src/meta/src/hummock/manager/compaction.rs

```diff
@@ -102,4 +116,149 @@ impl HummockManager {
         let ctx = dynamic_level_core.get_priority_levels(&levels, &status.level_handlers);
         ctx.score_levels
     }
+
+    pub async fn handle_pull_task_event(
+        &self,
+        context_id: u32,
+        pull_task_count: usize,
+        compaction_selectors: &mut HashMap<TaskType, Box<dyn CompactionSelector>>,
+    ) {
+        assert_ne!(0, pull_task_count);
+        if let Some(compactor) = self.compactor_manager.get_compactor(context_id) {
+            let (groups, task_type) = self.auto_pick_compaction_groups_and_type().await;
+            if !groups.is_empty() {
+                let selector: &mut Box<dyn CompactionSelector> =
+                    compaction_selectors.get_mut(&task_type).unwrap();
+
+                let mut generated_task_count = 0;
+                let mut existed_groups = groups.clone();
+                let mut meet_error = false;
+                let mut wait_compact_groups: HashSet<CompactionGroupId> = HashSet::default();
+
+                while generated_task_count < pull_task_count && !meet_error {
+                    let compact_ret = self
+                        .get_compact_tasks(std::mem::take(&mut existed_groups), selector)
+                        .await;
+
+                    match compact_ret {
+                        Ok((compact_tasks, trivial_tasks)) => {
+                            if compact_tasks.is_empty() {
+                                break;
+                            }
+                            generated_task_count += compact_tasks.len();
+                            for task in trivial_tasks {
+                                existed_groups.push(task.compaction_group_id);
+                                wait_compact_groups.insert(task.compaction_group_id);
+                            }
+                            for task in compact_tasks {
+                                let task_id = task.task_id;
+                                existed_groups.push(task.compaction_group_id);
+                                wait_compact_groups.insert(task.compaction_group_id);
+                                if let Err(e) =
+                                    compactor.send_event(ResponseEvent::CompactTask(task))
+                                {
+                                    tracing::warn!(
+                                        error = %e.as_report(),
+                                        "Failed to send task {} to {}",
+                                        task_id,
+                                        compactor.context_id(),
+                                    );
+                                    self.compactor_manager.remove_compactor(context_id);
+                                    meet_error = true;
+                                    break;
+                                }
+                            }
+                            existed_groups.sort();
+                            existed_groups.dedup();
```

Comment:
Should we randomize the order before going for the next round of `get_compact_tasks`?

Reply:
I think it is not necessary? Because we will pick at least one task for every group...

Reply:
No. Because we will check all groups in `get_compact_tasks `.

Reply:
> Because we will pick at least one task for every group...
> No. Because we will check all groups in get_compact_tasks .



After we introduced `max_select_count` in the latest commit, I think  this assumption is no longer true.